x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -185,30 +172,72 @@ func (rw *ReadWriter) WriteTraceEvent(traceID string, id string, event *modelpb.
 
 func (rw *ReadWriter) writeEntry(e *badger.Entry, opts WriterOpts) error {
 	rw.pendingWrites++
-	err := rw.txn.SetEntry(e.WithTTL(opts.TTL))
-	// Attempt to flush if there are 200 or more uncommitted writes.
-	// This ensures calls to ReadTraceEvents are not slowed down;
-	// ReadTraceEvents uses an iterator, which must sort all keys
-	// of uncommitted writes.
-	// The 200 value yielded a good balance between read and write speed:
-	// https://github.com/elastic/apm-server/pull/8407#issuecomment-1162994643
-	if rw.pendingWrites >= 200 {
-		if err := rw.Flush(opts.StorageLimitInBytes); err != nil {
+	entrySize := estimateSize(e)
+	lsm, vlog := rw.s.db.Size()
+
+	// there are multiple ReadWriters writing to the same storage so add
+	// the entry size and consider the new value to avoid TOCTOU issues.
+	pendingSize := rw.s.pendingSize.Add(entrySize)
+	rw.pendingSize += entrySize
+
+	if current := pendingSize + lsm + vlog; opts.StorageLimitInBytes != 0 && current >= opts.StorageLimitInBytes {
+		// Discard the txn and re-create it if the soft limit has been reached.
+		rw.txn.Discard()
```

Comment:
IIUC this is still discarding tx which are accepted (and returned as ok to the caller). I think we should flush the rw immediately instead. Also I'm thinking if we need to add a mutex somewhere to avoid race and overshooting the storage limit

Reply:
Flush will always recreate tx. Should be no need to discard and create new tx. Update the comment to say "Discard the entry"